src/main/java/net/neoforged/neoforge/fluids/crafting/UnitAmountFluidIngredient.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.fluids.crafting;
+
+import com.mojang.datafixers.util.Either;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import io.netty.buffer.ByteBuf;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.tags.TagKey;
+import net.minecraft.world.level.material.Fluid;
+import net.neoforged.neoforge.common.util.NeoForgeExtraCodecs;
+import net.neoforged.neoforge.fluids.FluidStack;
+import net.neoforged.neoforge.fluids.FluidUnit;
+import net.neoforged.neoforge.fluids.FluidUnitSpec;
+import net.neoforged.neoforge.fluids.FluidUnits;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Standard implementation for a FluidIngredient with an amount.
+ *
+ * <p>{@link FluidIngredient}, like its item counterpart, explicitly does not perform count checks,
+ * so this class is used to (a) wrap a standard FluidIngredient with an amount and (b) provide a
+ * standard serialization format for mods to use.
+ *
+ * @see net.neoforged.neoforge.common.crafting.SizedIngredient
+ */
+public final class UnitAmountFluidIngredient {
+    /**
+     * The codec for the "amount" field.
+     *
+     * <p>The amounts field consists of a list of maps of fluid units to integers.
+     * It can be simplified to a single map, or a quantity in millibuckets.
+     * For example, the following list specifies a recipe that takes 1 block and
+     * 3 ingots or 3 gems' worth of a molten metal or gem fluid.
+     *
+     * <pre>{@code
+     *  [{"c:block": 1, "c:ingot", 3},
+     *   {"c:block": 1, "c:gem", 3}]
+     * }</pre>
+     *
+     * <p>If none of the entries has all of its units defined in the fluid's unit spec,
+     * the ingredient will not match the fluid.
+     *
+     */
+    private static final Codec<List<Map<FluidUnit, Integer>>> AMOUNTS_CODEC = makeAmountsCodec();
```

Comment:
Can we make this public? This might be useful in some contexts where you don't have a specific fluid but you do have a specific amount.